### ISSUE TITLE
Wrong compiler error message.

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#67623]
```rust
trait Baz {}
trait Bar2<A> { type X; }
struct Foo<'a>(&'a str);
impl<'a, A, B> Baz for B where B: Bar2<Foo<'a>, X=A> {}
```
The current error message is 
"error[E0207]: the type parameter `A` is not constrained by the impl trait, self type, or predicates"

but the actually error in the code is that `'a` is not constrained. `A` is used in associated type, and is correct according to the documentation of `E0207`. The correct fix would be move to `'a` to a HRTB as in `for<'a> Bar2<Foo<'a>, X=A>`

**Comment by Globidev** [comment]
Is `'a` actually unconstrained ? If we follow the same logic as type parameters then it should be, but that currently compiles fine without any uncosntrained error:
```rs
impl<'a, B> Baz for B where B: Bar2<Foo<'a>> {}
```

**Comment by crlf0710** [comment]
According to E0207: `rustc --explain E0207`
```text
Any type parameter or lifetime parameter of an `impl` must meet at least one of
the following criteria:

 - it appears in the _implementing type_ of the impl, e.g. `impl<T> Foo<T>`
 - for a trait impl, it appears in the _implemented trait_, e.g.
   `impl<T> SomeTrait<T> for Foo`
 - it is bound as an associated type, e.g. `impl<T, U> SomeTrait for T
   where T: AnotherTrait<AssocType=U>`
```

**Comment by estebank** [comment]
Triage: no change.
